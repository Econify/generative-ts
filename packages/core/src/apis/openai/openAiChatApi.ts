import * as t from "io-ts";
import type { TypeOf } from "io-ts";
import { isLeft } from "fp-ts/Either";

import type { ModelApi, ModelRequestOptions } from "@typeDefs";

import { EjsTemplate } from "../../utils/ejsTemplate";

import { nullable } from "../_utils/io-ts-nullable";

import type { FewShotRequestOptions } from "../shared/fewShot";

const templateSource = `{
  "model": "<%= modelId %>",
  "messages": [
    <% if (typeof system !== 'undefined') { %>
    {
      "role": "system",
      "content": "<%= system %>"
    },
    <% } %>
    <% (typeof examplePairs !== 'undefined' ? examplePairs : []).forEach(pair => { %>
    {
      "role": "user",
      "content": "<%= pair.user %>"
    },
    {
      "role": "assistant",
      "content": "<%= pair.assistant %>"
    },
    <% }) %>
    <% (typeof messages !== 'undefined' ? messages : []).forEach(message => { %>
    {
      "role": "<%= message.role %>",
      "content": "<%= message.content %>"
    },
    <% }) %>
    {
      "role": "user",
      "content": "<%= prompt %>"
    }
  ]
  <% if (typeof frequency_penalty !== 'undefined') { %>, "frequency_penalty": <%= frequency_penalty %><% } %>
  <% if (typeof logit_bias !== 'undefined') { %>, "logit_bias": <%- JSON.stringify(logit_bias) %><% } %>
  <% if (typeof logprobs !== 'undefined') { %>, "logprobs": <%= logprobs %><% } %>
  <% if (typeof top_logprobs !== 'undefined') { %>, "top_logprobs": <%= top_logprobs %><% } %>
  <% if (typeof max_tokens !== 'undefined') { %>, "max_tokens": <%= max_tokens %><% } %>
  <% if (typeof n !== 'undefined') { %>, "n": <%= n %><% } %>
  <% if (typeof presence_penalty !== 'undefined') { %>, "presence_penalty": <%= presence_penalty %><% } %>
  <% if (typeof response_format !== 'undefined') { %>, "response_format": <%- JSON.stringify(response_format) %><% } %>
  <% if (typeof seed !== 'undefined') { %>, "seed": <%= seed %><% } %>
  <% if (typeof stop !== 'undefined') { %>, "stop": <%- JSON.stringify(stop) %><% } %>
  <% if (typeof stream !== 'undefined') { %>, "stream": <%= stream %><% } %>
  <% if (typeof stream_options !== 'undefined') { %>, "stream_options": <%- JSON.stringify(stream_options) %><% } %>
  <% if (typeof temperature !== 'undefined') { %>, "temperature": <%= temperature %><% } %>
  <% if (typeof top_p !== 'undefined') { %>, "top_p": <%= top_p %><% } %>
  <% if (typeof user !== 'undefined') { %>, "user": "<%= user %>"<% } %>
  <% if (typeof tools !== 'undefined') { %>, "tools": <%- JSON.stringify(tools) %><% } %>
  <% if (typeof tool_choice !== 'undefined') { %>, "tool_choice": <%- JSON.stringify(tool_choice) %><% } %>
  <% if (typeof function_call !== 'undefined') { %>, "function_call": <%= function_call %><% } %>
  <% if (typeof functions !== 'undefined') { %>, "functions": <%- JSON.stringify(functions) %><% } %>
}`;

/**
 * @category OpenAI ChatCompletion
 * @category Requests
 */
export interface OpenAiChatOptions
  extends ModelRequestOptions,
    FewShotRequestOptions {
  messages?: {
    role: "user" | "assistant" | "system";
    content: string;
  }[];
  frequency_penalty?: number;
  logit_bias?: Record<string, number>;
  logprobs?: boolean;
  top_logprobs?: number;
  max_tokens?: number;
  n?: number;
  presence_penalty?: number;
  response_format?: {
    type: "text" | "json_object";
  };
  seed?: number;
  stop?: string | string[];
  stream?: boolean;
  stream_options?: {
    include_usage: boolean;
  };
  temperature?: number;
  top_p?: number;
  user?: string;
  tools?: {
    type: "function";
    function: {
      name: string;
      description?: string;
      parameters?: object; // TODO: JsonSchema
    };
  }[];
  tool_choice?:
    | "none"
    | "auto"
    | "required"
    | {
        type: "function";
        function: {
          name: string;
        };
      };
  function_call?: string; // "none" | "auto"
  functions?: {
    name: string;
    description?: string;
    parameters?: object; // TODO: JsonSchema
  }[];

  // LMStudio apparently has these additional options:
  // top_k
  // repeat_penalty
}

/**
 * @category OpenAI ChatCompletion
 * @category Templates
 */
export const OpenAiChatTemplate = new EjsTemplate<OpenAiChatOptions>(
  templateSource,
);

const ChatCompletionResponseMessage = t.intersection([
  t.type({
    role: t.string,
    content: t.string, // TODO nullable(t.string) ??
  }),
  t.partial({
    tool_calls: t.array(
      t.type({
        id: t.string,
        type: t.string, // will always be "function"
        function: t.type({
          name: t.string,
          arguments: t.string,
        }),
      }),
    ),
    function_call: t.type({
      name: t.string,
      arguments: t.string,
    }),
  }),
]);

const OpenAiChatResponseCodec = t.intersection([
  t.type({
    id: t.string,
    model: t.string,
    object: t.string,
    created: t.number,
    choices: t.array(
      t.type({
        finish_reason: t.string,
        index: t.number,
        message: ChatCompletionResponseMessage,
        logprobs: nullable(
          t.type({
            content: nullable(
              t.array(
                t.type({
                  token: t.string,
                  logprob: t.number,
                  bytes: nullable(t.array(t.number)),
                  top_logprobs: t.array(
                    t.type({
                      token: t.string,
                      logprob: t.number,
                      bytes: nullable(t.array(t.number)),
                    }),
                  ),
                }),
              ),
            ),
          }),
        ),
      }),
    ),
  }),
  t.partial({
    system_fingerprint: t.string,
    usage: t.type({
      completion_tokens: t.number,
      prompt_tokens: t.number,
      total_tokens: t.number,
    }),
  }),
]);

/**
 * @category OpenAI ChatCompletion
 * @category Responses
 */
export interface OpenAiChatResponse
  extends TypeOf<typeof OpenAiChatResponseCodec> {}

export function isOpenAiChatResponse(
  response: unknown,
): response is OpenAiChatResponse {
  return !isLeft(OpenAiChatResponseCodec.decode(response));
}

/**
 *
 * ## Reference
 * [OpenAI Chat Completion](https://github.com/openai/openai-openapi/)
 *
 * ## Providers using this API
 * - {@link createOpenAiChatModelProvider | OpenAI}
 * - {@link createGroqModelProvider | Groq}
 * - {@link createLmStudioModelProvider | LMStudio}
 *
 * @category APIs
 * @category OpenAI ChatCompletion
 * @category Provider: OpenAI
 * @category Provider: Groq
 * @category Provider: LMStudio
 *
 */
export const OpenAiChatApi: ModelApi<OpenAiChatOptions, OpenAiChatResponse> = {
  requestTemplate: OpenAiChatTemplate,
  responseGuard: isOpenAiChatResponse,
};
