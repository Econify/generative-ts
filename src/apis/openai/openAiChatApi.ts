import * as t from "io-ts";
import type { TypeOf } from "io-ts";
import { isLeft } from "fp-ts/Either";

import type { ModelApi, ModelRequestOptions } from "../../typeDefs";

import { Template } from "../../utils/template";

import type { FewShotRequestOptions } from "../_shared_interfaces/fewShot";

const templateSource = `{
  "model": "{{ modelId }}",
  "messages": [
    {% if system %}
    {
      "role": "system",
      "content": "{{ system | safe }}"
    },
    {% endif %}
    {% for pair in examplePairs %}
    {
      "role": "user",
      "content": "{{ pair.user | safe }}"
    },
    {
      "role": "assistant",
      "content": "{{ pair.assistant | safe }}"
    },
    {% endfor %}
    {
      "role": "user",
      "content": "{{ prompt | safe }}"
    }
  ]
  {% if temperature %}, "temperature": {{ temperature }}{% endif %}
  {% if max_tokens %}, "max_tokens": {{ max_tokens }}{% endif %}
  {% if frequency_penalty %}, "frequency_penalty": {{ frequency_penalty }}{% endif %}
  {% if logprobs %}, "logprobs": true{% endif %}
  {% if top_logprobs %}, "top_logprobs": {{ top_logprobs }}{% endif %}
  {% if n %}, "n": {{ n }}{% endif %}
  {% if presence_penalty %}, "presence_penalty": {{ presence_penalty }}{% endif %}
  {% if seed %}, "seed": {{ seed }}{% endif %}
  {% if stop %}, "stop": {{ stop | to_json }}{% endif %}
  {% if top_p %}, "top_p": {{ top_p }}{% endif %}
}`;

export interface OpenAiChatOptions
  extends ModelRequestOptions,
    FewShotRequestOptions {
  temperature?: number;
  max_tokens?: number;
  frequency_penalty?: number;
  // logit_bias: Record<string, number>;
  logprobs?: boolean;
  top_logprobs?: number;
  n?: number;
  presence_penalty?: number;
  // response_format
  seed?: number;
  stop?: string | string[];
  // stream
  // stream_options
  top_p?: number;
  // tools
  // tool_choice
  // user
  // function_call
  // functions
}

export const OpenAiChatTemplate = new Template<OpenAiChatOptions>(
  templateSource,
);

const OpenAiChatResponseCodec = t.type({
  id: t.string,
  object: t.string,
  created: t.number,
  model: t.string,
  choices: t.array(
    t.type({
      index: t.number,
      finish_reason: t.string,
      message: t.type({
        role: t.string,
        content: t.string,
        // tool_calls
        // function_call
      }),
      // logprobs: t.type({
      //   content: t.array(
      //     t.type({
      //       token: t.string,
      //       logprob: t.number,
      //       // bytes
      //       // top_logprobs
      //     }),
      //   ),
      // }),
    }),
  ),
  usage: t.type({
    prompt_tokens: t.number,
    completion_tokens: t.number,
    total_tokens: t.number,
  }),
  // system_fingerprint: t.string,
});

export type OpenAiChatResponse = TypeOf<typeof OpenAiChatResponseCodec>;

export function isOpenAiChatResponse(
  response: unknown,
): response is OpenAiChatResponse {
  return !isLeft(OpenAiChatResponseCodec.decode(response));
}

export const OpenAiChatApi: ModelApi<OpenAiChatOptions, OpenAiChatResponse> = {
  requestTemplate: OpenAiChatTemplate,
  responseGuard: isOpenAiChatResponse,
};
