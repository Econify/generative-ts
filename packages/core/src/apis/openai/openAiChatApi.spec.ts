import { OpenAiChatApi, OpenAiChatOptions } from "./openAiChatApi";

function render(context: Omit<OpenAiChatOptions, "modelId">) {
  try {
    const rendered = OpenAiChatApi.requestTemplate.render({
      modelId: "mock-model-id",
      ...context,
    });
    return JSON.parse(rendered);
  } catch (error) {
    throw new Error("Invalid JSON output");
  }
}

describe("OpenAiChatApi:", () => {
  /**
   * FewShotRequestOptions (prompt, examplePairs, system):
   */

  test("$prompt", () => {
    const rendered = render({
      $prompt: "mock-prompt",
    });

    expect(rendered).toMatchSnapshot();
  });

  test("$prompt, examplePairs", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      examplePairs: [
        { user: "mock-user-msg-1", assistant: "mock-assistant-msg-1" },
        { user: "mock-user-msg-2", assistant: "mock-assistant-msg-2" },
      ],
    });

    expect(rendered).toMatchSnapshot();
  });

  test("$prompt, system", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      system: "mock-system-text",
    });

    expect(rendered).toMatchSnapshot();
  });

  test("$prompt, examplePairs, system", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      examplePairs: [
        { user: "mock-user-msg-1", assistant: "mock-assistant-msg-1" },
      ],
      system: "mock-system-text",
    });

    expect(rendered).toMatchSnapshot();
  });

  /**
   * "Native" few shot options (messages):
   */

  test("$prompt, messages", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      messages: [
        {
          role: "assistant",
          content: "mock-assistant-text",
        },
        {
          role: "user",
          content: "mock-user-text",
        },
        {
          role: "assistant",
          content: "mock-assistant-text-2",
        },
      ],
    });

    expect(rendered).toMatchSnapshot();
  });

  test("$prompt, examplePairs, messages", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      examplePairs: [
        { user: "mock-user-msg-1", assistant: "mock-assistant-msg-1" },
      ],
      messages: [
        {
          role: "assistant",
          content: "mock-assistant-text",
        },
      ],
    });

    expect(rendered).toMatchSnapshot();
  });

  test("$prompt, system, messages", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      system: "mock-system-text",
      messages: [
        {
          role: "assistant",
          content: "mock-assistant-text",
        },
        {
          role: "user",
          content: "mock-user-text",
        },
        {
          role: "assistant",
          content: "mock-assistant-text-2",
        },
      ],
    });

    expect(rendered).toMatchSnapshot();
  });

  test("$prompt, examplePairs, system, messages", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      examplePairs: [
        { user: "mock-user-msg-1", assistant: "mock-assistant-msg-1" },
      ],
      system: "mock-system-text",
      messages: [
        {
          role: "assistant",
          content: "mock-assistant-text",
        },
      ],
    });

    expect(rendered).toMatchSnapshot();
  });

  /**
   * Tool-related:
   */

  test("$prompt, messages with function_call", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      messages: [
        {
          role: "assistant",
          content: "mock-assistant-text",
          function_call: {
            name: "mock-function",
            arguments: '{ "key": "value" }',
          },
        },
      ],
    });

    expect(rendered).toMatchSnapshot();
  });

  test("$prompt, tools", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      tools: [
        {
          type: "function",
          function: {
            name: "mock-function",
            description: "mock-description",
            parameters: {
              type: "object",
              properties: {
                key: { type: "string" },
              },
            },
          },
        },
      ],
    });

    expect(rendered).toMatchSnapshot();
  });

  test("$prompt, tool_choice", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      tool_choice: {
        type: "function",
        function: {
          name: "mock-function",
        },
      },
    });

    expect(rendered).toMatchSnapshot();
  });

  test("$prompt, functions", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      functions: [
        {
          name: "mock-function",
          description: "mock-description",
          parameters: {
            type: "object",
            properties: {
              key: { type: "string" },
            },
          },
        },
      ],
    });

    expect(rendered).toMatchSnapshot();
  });

  /**
   * All options:
   */

  test("all options", () => {
    const rendered = render({
      $prompt: "mock-prompt",
      examplePairs: [
        { user: "mock-user-msg-1", assistant: "mock-assistant-msg-1" },
      ],
      messages: [
        {
          role: "assistant",
          content: "mock-assistant-text",
          function_call: {
            name: "mock-function",
            arguments: '{ "key": "value" }',
          },
        },
      ],
      system: "mock-system-text",
      frequency_penalty: 0.5,
      logit_bias: {
        "50256": -100,
      },
      logprobs: true,
      top_logprobs: 5,
      max_tokens: 1000,
      n: 3,
      presence_penalty: 0.3,
      response_format: {
        type: "json_object",
      },
      seed: 1234,
      stop: ["mock-stop"],
      stream: true,
      stream_options: {
        include_usage: true,
      },
      temperature: 0.7,
      top_p: 0.9,
      user: "mock-user",
      tools: [
        {
          type: "function",
          function: {
            name: "mock-tool-function",
            description: "mock-description",
            parameters: {
              type: "object",
              properties: {
                key: { type: "string" },
              },
            },
          },
        },
      ],
      tool_choice: {
        type: "function",
        function: {
          name: "mock-tool-choice-function",
        },
      },
      function_call: "auto",
      functions: [
        {
          name: "mock-function",
          description: "mock-description",
          parameters: {
            type: "object",
            properties: {
              key: { type: "string" },
            },
          },
        },
      ],
    });

    expect(rendered).toMatchSnapshot();
  });
});
