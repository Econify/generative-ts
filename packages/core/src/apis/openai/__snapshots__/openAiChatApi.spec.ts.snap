// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenAiChatApi: $prompt 1`] = `
{
  "messages": [
    {
      "content": "mock-prompt",
      "role": "user",
    },
  ],
  "model": "mock-model-id",
}
`;

exports[`OpenAiChatApi: $prompt, examplePairs 1`] = `
{
  "messages": [
    {
      "content": "mock-user-msg-1",
      "role": "user",
    },
    {
      "content": "mock-assistant-msg-1",
      "role": "assistant",
    },
    {
      "content": "mock-user-msg-2",
      "role": "user",
    },
    {
      "content": "mock-assistant-msg-2",
      "role": "assistant",
    },
    {
      "content": "mock-prompt",
      "role": "user",
    },
  ],
  "model": "mock-model-id",
}
`;

exports[`OpenAiChatApi: $prompt, examplePairs, messages 1`] = `
{
  "messages": [
    {
      "content": "mock-user-msg-1",
      "role": "user",
    },
    {
      "content": "mock-assistant-msg-1",
      "role": "assistant",
    },
    {
      "content": "mock-assistant-text",
      "role": "assistant",
    },
    {
      "content": "mock-prompt",
      "role": "user",
    },
  ],
  "model": "mock-model-id",
}
`;

exports[`OpenAiChatApi: $prompt, examplePairs, system 1`] = `
{
  "messages": [
    {
      "content": "mock-system-text",
      "role": "system",
    },
    {
      "content": "mock-user-msg-1",
      "role": "user",
    },
    {
      "content": "mock-assistant-msg-1",
      "role": "assistant",
    },
    {
      "content": "mock-prompt",
      "role": "user",
    },
  ],
  "model": "mock-model-id",
}
`;

exports[`OpenAiChatApi: $prompt, examplePairs, system, messages 1`] = `
{
  "messages": [
    {
      "content": "mock-system-text",
      "role": "system",
    },
    {
      "content": "mock-user-msg-1",
      "role": "user",
    },
    {
      "content": "mock-assistant-msg-1",
      "role": "assistant",
    },
    {
      "content": "mock-assistant-text",
      "role": "assistant",
    },
    {
      "content": "mock-prompt",
      "role": "user",
    },
  ],
  "model": "mock-model-id",
}
`;

exports[`OpenAiChatApi: $prompt, functions 1`] = `
{
  "functions": [
    {
      "description": "mock-description",
      "name": "mock-function",
      "parameters": {
        "properties": {
          "key": {
            "type": "string",
          },
        },
        "type": "object",
      },
    },
  ],
  "messages": [
    {
      "content": "mock-prompt",
      "role": "user",
    },
  ],
  "model": "mock-model-id",
}
`;

exports[`OpenAiChatApi: $prompt, messages 1`] = `
{
  "messages": [
    {
      "content": "mock-assistant-text",
      "role": "assistant",
    },
    {
      "content": "mock-user-text",
      "role": "user",
    },
    {
      "content": "mock-assistant-text-2",
      "role": "assistant",
    },
    {
      "content": "mock-prompt",
      "role": "user",
    },
  ],
  "model": "mock-model-id",
}
`;

exports[`OpenAiChatApi: $prompt, messages with function_call 1`] = `
{
  "messages": [
    {
      "content": "mock-assistant-text",
      "function_call": {
        "arguments": "{ "key": "value" }",
        "name": "mock-function",
      },
      "role": "assistant",
    },
    {
      "content": "mock-prompt",
      "role": "user",
    },
  ],
  "model": "mock-model-id",
}
`;

exports[`OpenAiChatApi: $prompt, system 1`] = `
{
  "messages": [
    {
      "content": "mock-system-text",
      "role": "system",
    },
    {
      "content": "mock-prompt",
      "role": "user",
    },
  ],
  "model": "mock-model-id",
}
`;

exports[`OpenAiChatApi: $prompt, system, messages 1`] = `
{
  "messages": [
    {
      "content": "mock-system-text",
      "role": "system",
    },
    {
      "content": "mock-assistant-text",
      "role": "assistant",
    },
    {
      "content": "mock-user-text",
      "role": "user",
    },
    {
      "content": "mock-assistant-text-2",
      "role": "assistant",
    },
    {
      "content": "mock-prompt",
      "role": "user",
    },
  ],
  "model": "mock-model-id",
}
`;

exports[`OpenAiChatApi: $prompt, tool_choice 1`] = `
{
  "messages": [
    {
      "content": "mock-prompt",
      "role": "user",
    },
  ],
  "model": "mock-model-id",
  "tool_choice": {
    "function": {
      "name": "mock-function",
    },
    "type": "function",
  },
}
`;

exports[`OpenAiChatApi: $prompt, tools 1`] = `
{
  "messages": [
    {
      "content": "mock-prompt",
      "role": "user",
    },
  ],
  "model": "mock-model-id",
  "tools": [
    {
      "function": {
        "description": "mock-description",
        "name": "mock-function",
        "parameters": {
          "properties": {
            "key": {
              "type": "string",
            },
          },
          "type": "object",
        },
      },
      "type": "function",
    },
  ],
}
`;

exports[`OpenAiChatApi: all options 1`] = `
{
  "frequency_penalty": 0.5,
  "function_call": "auto",
  "functions": [
    {
      "description": "mock-description",
      "name": "mock-function",
      "parameters": {
        "properties": {
          "key": {
            "type": "string",
          },
        },
        "type": "object",
      },
    },
  ],
  "logit_bias": {
    "50256": -100,
  },
  "logprobs": true,
  "max_tokens": 1000,
  "messages": [
    {
      "content": "mock-system-text",
      "role": "system",
    },
    {
      "content": "mock-user-msg-1",
      "role": "user",
    },
    {
      "content": "mock-assistant-msg-1",
      "role": "assistant",
    },
    {
      "content": "mock-assistant-text",
      "function_call": {
        "arguments": "{ "key": "value" }",
        "name": "mock-function",
      },
      "role": "assistant",
    },
    {
      "content": "mock-prompt",
      "role": "user",
    },
  ],
  "model": "mock-model-id",
  "n": 3,
  "presence_penalty": 0.3,
  "response_format": {
    "type": "json_object",
  },
  "seed": 1234,
  "stop": [
    "mock-stop",
  ],
  "stream": true,
  "stream_options": {
    "include_usage": true,
  },
  "temperature": 0.7,
  "tool_choice": {
    "function": {
      "name": "mock-tool-choice-function",
    },
    "type": "function",
  },
  "tools": [
    {
      "function": {
        "description": "mock-description",
        "name": "mock-tool-function",
        "parameters": {
          "properties": {
            "key": {
              "type": "string",
            },
          },
          "type": "object",
        },
      },
      "type": "function",
    },
  ],
  "top_logprobs": 5,
  "top_p": 0.9,
  "user": "mock-user",
}
`;
